File: _code-samples/send-a-trust-line-token/py/send_trust_line_token.py
import asyncio
import json
import os
import sys

from xrpl.clients import JsonRpcClient
from xrpl.models import AccountLines
from xrpl.models.amounts import IssuedCurrencyAmount
from xrpl.models.transactions import Payment, TrustSet
from xrpl.transaction import submit_and_wait
from xrpl.wallet import Wallet

from send_trust_line_token_setup import main as run_setup

# Set up client ----------------------
client = JsonRpcClient("https://s.altnet.rippletest.net:51234")

# This step checks for the necessary setup data to run the tutorial.
# If missing, send_trust_line_token_setup.py will generate it.
if not os.path.exists("send_trust_line_token_setup.json"):
    print("\n=== Tutorial setup data doesn't exist. Running setup script... ===\n")
    asyncio.run(run_setup())

# Load preconfigured issuer, sender, and receiver accounts.
with open("send_trust_line_token_setup.json") as f:
    setup_data = json.load(f)

issuer = Wallet.from_seed(setup_data["issuer"]["seed"])
sender = Wallet.from_seed(setup_data["sender"]["seed"])
receiver = Wallet.from_seed(setup_data["receiver"]["seed"])
currency_code = "FOO"

print(f"Issuer address:   {issuer.address}")
print(f"Sender address:   {sender.address}")
print(f"Receiver address: {receiver.address}")

# Create trust line ----------------------
# The receiver opts in to the token by creating a trust line to the issuer.
# The limit_amount sets the maximum amount of the token the receiver will hold.
print("\n=== Creating trust line from receiver to issuer... ===\n")
trust_set_tx = TrustSet(
    account=receiver.address,
    limit_amount=IssuedCurrencyAmount(
        currency=currency_code,
        issuer=issuer.address,
        value="1000000000",
    ),
)
print(json.dumps(trust_set_tx.to_xrpl(), indent=2))

trust_set_response = submit_and_wait(trust_set_tx, client, receiver)
result_code = trust_set_response.result["meta"]["TransactionResult"]
if result_code != "tesSUCCESS":
    print(f"Error: Unable to create trust line: {result_code}")
    sys.exit(1)
print("Trust line created from receiver to issuer!")
print(f"Explorer link: https://testnet.xrpl.org/transactions/{trust_set_response.result['hash']}")

# Check initial balances ----------------------
# get_trust_line_balance returns the trust line balance between `account` and
# `peer` for the given currency.
def get_trust_line_balance(account, currency, peer):
    response = client.request(AccountLines(
        account=account,
        peer=peer,
        ledger_index="validated",
    ))
    for line in response.result.get("lines", []):
        if line["currency"] == currency:
            return line["balance"]
    return "0"

print(f"\n=== Checking initial {currency_code} balances... ===\n")
print("Holders' perspective:")
print(f"  Sender's balance:   {get_trust_line_balance(sender.address, currency_code, issuer.address)}")
print(f"  Receiver's balance: {get_trust_line_balance(receiver.address, currency_code, issuer.address)}")
print("Issuer's perspective:")
print(f"  Owed to sender:   {get_trust_line_balance(issuer.address, currency_code, sender.address)}")
print(f"  Owed to receiver: {get_trust_line_balance(issuer.address, currency_code, receiver.address)}")

# Send issued token ----------------------
# The sender pays the receiver with the issued currency.
send_quantity = "100"
print(f"\n=== Sending {currency_code} payment... ===\n")
payment_tx = Payment(
    account=sender.address,
    destination=receiver.address,
    amount=IssuedCurrencyAmount(
        currency=currency_code,
        issuer=issuer.address,
        value=send_quantity,
    ),
)
print(json.dumps(payment_tx.to_xrpl(), indent=2))

payment_response = submit_and_wait(payment_tx, client, sender)
result_code = payment_response.result["meta"]["TransactionResult"]
if result_code != "tesSUCCESS":
    print(f"Error: Unable to send token: {result_code}")
    sys.exit(1)
print("Payment successful!")
print(f"Explorer link: https://testnet.xrpl.org/transactions/{payment_response.result['hash']}")

# Verify balances ----------------------
# Check account balances after the payment.
print(f"\n=== Checking final {currency_code} balances... ===\n")
print("Holders' perspective:")
print(f"  Sender's balance:   {get_trust_line_balance(sender.address, currency_code, issuer.address)}")
print(f"  Receiver's balance: {get_trust_line_balance(receiver.address, currency_code, issuer.address)}")
print("Issuer's perspective:")
print(f"  Owed to sender:   {get_trust_line_balance(issuer.address, currency_code, sender.address)}")
print(f"  Owed to receiver: {get_trust_line_balance(issuer.address, currency_code, receiver.address)}")

File: _code-samples/send-a-trust-line-token/js/sendTrustLineToken.js
import xrpl from 'xrpl'
import fs from 'fs'
import { setup } from './sendTrustLineTokenSetup.js'

// Set up client ----------------------
const client = new xrpl.Client('wss://s.altnet.rippletest.net:51233')
await client.connect()

// This step checks for the necessary setup data to run the tutorial.
// If missing, sendTrustLineTokenSetup.js will generate it.
if (!fs.existsSync('sendTrustLineTokenSetup.json')) {
  console.log(`\n=== Tutorial setup data doesn't exist. Running setup script... ===\n`)
  await setup()
}

// Load preconfigured issuer, sender, and receiver accounts.
const setupData = JSON.parse(fs.readFileSync('sendTrustLineTokenSetup.json', 'utf8'))
const issuer = xrpl.Wallet.fromSeed(setupData.issuer.seed)
const sender = xrpl.Wallet.fromSeed(setupData.sender.seed)
const receiver = xrpl.Wallet.fromSeed(setupData.receiver.seed)
const currencyCode = 'FOO'

console.log(`Issuer address:   ${issuer.address}`)
console.log(`Sender address:   ${sender.address}`)
console.log(`Receiver address: ${receiver.address}`)

// Create trust line ----------------------
// The receiver opts in to the token by creating a trust line to the issuer.
// The LimitAmount sets the maximum amount of the token the receiver will hold.
console.log(`\n=== Creating trust line from receiver to issuer... ===\n`)
const trustSetTx = {
  TransactionType: 'TrustSet',
  Account: receiver.address,
  LimitAmount: {
    currency: currencyCode,
    issuer: issuer.address,
    value: '1000000000'
  }
}
xrpl.validate(trustSetTx)
console.log(JSON.stringify(trustSetTx, null, 2))

const trustSetResponse = await client.submitAndWait(trustSetTx, {
  wallet: receiver,
  autofill: true
})
if (trustSetResponse.result.meta?.TransactionResult !== 'tesSUCCESS') {
  const resultCode = trustSetResponse.result.meta?.TransactionResult
  console.error('Error: Unable to create trust line:', resultCode)
  await client.disconnect()
  process.exit(1)
}
console.log('Trust line created from receiver to issuer!')
console.log(`Explorer link: https://testnet.xrpl.org/transactions/${trustSetResponse.result.hash}`)

// Check initial balances ----------------------
// getTrustLineBalance returns the trust line balance between `account` and
// `peer` for the given currency.
async function getTrustLineBalance(account, currency, peer) {
  const response = await client.request({
    command: 'account_lines',
    account,
    peer,
    ledger_index: 'validated'
  })
  const line = response.result.lines.find(l => l.currency === currency)
  return line?.balance ?? '0'
}

console.log(`\n=== Checking initial ${currencyCode} balances... ===\n`)
console.log("Holders' perspective:")
console.log(`  Sender's balance:   ${await getTrustLineBalance(sender.address, currencyCode, issuer.address)}`)
console.log(`  Receiver's balance: ${await getTrustLineBalance(receiver.address, currencyCode, issuer.address)}`)
console.log("Issuer's perspective:")
console.log(`  Owed to sender:   ${await getTrustLineBalance(issuer.address, currencyCode, sender.address)}`)
console.log(`  Owed to receiver: ${await getTrustLineBalance(issuer.address, currencyCode, receiver.address)}`)

// Send issued token ----------------------
// The sender pays the receiver with the issued currency.
const sendQuantity = '100'
console.log(`\n=== Sending ${currencyCode} payment... ===\n`)
const paymentTx = {
  TransactionType: 'Payment',
  Account: sender.address,
  Destination: receiver.address,
  Amount: {
    currency: currencyCode,
    issuer: issuer.address,
    value: sendQuantity
  }
}
xrpl.validate(paymentTx)
console.log(JSON.stringify(paymentTx, null, 2))

const paymentResponse = await client.submitAndWait(paymentTx, {
  wallet: sender,
  autofill: true
})
if (paymentResponse.result.meta?.TransactionResult !== 'tesSUCCESS') {
  const resultCode = paymentResponse.result.meta?.TransactionResult
  console.error('Error: Unable to send token:', resultCode)
  await client.disconnect()
  process.exit(1)
}
console.log('Payment successful!')
console.log(`Explorer link: https://testnet.xrpl.org/transactions/${paymentResponse.result.hash}`)

// Verify balances ----------------------
// Check account balances after the payment.
console.log(`\n=== Checking final ${currencyCode} balances... ===\n`)
console.log("Holders' perspective:")
console.log(`  Sender's balance:   ${await getTrustLineBalance(sender.address, currencyCode, issuer.address)}`)
console.log(`  Receiver's balance: ${await getTrustLineBalance(receiver.address, currencyCode, issuer.address)}`)
console.log("Issuer's perspective:")
console.log(`  Owed to sender:   ${await getTrustLineBalance(issuer.address, currencyCode, sender.address)}`)
console.log(`  Owed to receiver: ${await getTrustLineBalance(issuer.address, currencyCode, receiver.address)}`)

await client.disconnect()
